docs/topics/async.txt

```diff
@@ -63,81 +63,88 @@ deploy Django using :doc:`ASGI </howto/deployment/asgi/index>` instead.
-on your code. In some cases, there may be a performance increase even for
+then back to async mode for the view. Django will also hold the sync thread
+open for middleware exception propagation. This may not be noticeable at first,
+but adding this penalty of one thread per request can remove any async
+performance advantage.
+
+You should do your own performance testing to see what effect ASGI versus WSGI
+has on your code. In some cases, there may be a performance increase even for
 purely-synchronous codebase under ASGI because the request-handling code is
-still all running asynchronously. In general, though, you will only want to
-enable ASGI mode if you have asynchronous code in your site.
+still all running asynchronously. In general you will only want to enable ASGI
+mode if you have asynchronous code in your project.
 
 .. _async-safety:
 
-Async-safety
+Async safety
 ============
 
-Certain key parts of Django are not able to operate safely in an asynchronous
+Certain key parts of Django are not able to operate safely in an async
 environment, as they have global state that is not coroutine-aware. These parts
 of Django are classified as "async-unsafe", and are protected from execution in
-an asynchronous environment. The ORM is the main example, but there are other
-parts that are also protected in this way.
+an async environment. The ORM is the main example, but there are other parts
+that are also protected in this way.
 
 If you try to run any of these parts from a thread where there is a *running
 event loop*, you will get a
 :exc:`~django.core.exceptions.SynchronousOnlyOperation` error. Note that you
 don't have to be inside an async function directly to have this error occur. If
-you have called a synchronous function directly from an asynchronous function
-without going through something like :func:`sync_to_async` or a threadpool,
-then it can also occur, as your code is still running in an asynchronous
-context.
+you have called a sync function directly from an async function,
+without using :func:`sync_to_async` or similar, then it can also occur. This is
+because as your code is still running asynchronously.
 
 If you encounter this error, you should fix your code to not call the offending
-code from an async context; instead, write your code that talks to async-unsafe
-in its own, synchronous function, and call that using
-:func:`asgiref.sync.sync_to_async`, or any other preferred way of running
-synchronous code in its own thread.
-
-If you are *absolutely* in dire need to run this code from an asynchronous
-context - for example, it is being forced on you by an external environment,
-and you are sure there is no chance of it being run concurrently (e.g. you are
-in a Jupyter_ notebook), then you can disable the warning with the
-``DJANGO_ALLOW_ASYNC_UNSAFE`` environment variable.
+code from an async context. Instead, write your code that talks to async-unsafe
+in its own, sync function, and call that using
```

Comment:
This needs a "code" or "APIs" or _something_ after async-unsafe

Reply:
Yes. Also bracketing the (or any other) since it's a bit of extra info.